webapp/channels/src/utils/performance_telemetry/reporter.ts

```diff
@@ -0,0 +1,277 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import type {Store} from 'redux';
+import {onCLS, onFCP, onINP, onLCP, onTTFB} from 'web-vitals';
+import type {Metric} from 'web-vitals';
+
+import type {Client4} from '@mattermost/client';
+
+import {getConfig} from 'mattermost-redux/selectors/entities/general';
+import {getCurrentUserId} from 'mattermost-redux/selectors/entities/users';
+
+import type {GlobalState} from 'types/store';
+
+import type {PerformanceLongTaskTiming} from './long_task';
+
+type PerformanceReportMeasure = {
+    metric: string;
+    value: number;
+    timestamp: number;
+}
+
+type PerformanceReport = {
+
+    // TODO confirm the version number that we want to start with
+    version: '0';
+
+    // TODO client ID?
+
+    platform: string;
+    user_agent: string;
+
+    start: number;
+    end: number;
+
+    counters: PerformanceReportMeasure[];
+    histograms: PerformanceReportMeasure[];
+}
+
+export default class PerformanceReporter {
+    private client: Client4;
+    private store: Store<GlobalState>;
+
+    private counters: Map<string, number>;
+    private histogramMeasures: PerformanceReportMeasure[];
+
+    private observer: PerformanceObserver;
+    private reportTimeout: number | undefined;
+
+    protected reportPeriodBase = 60 * 1000;
+    protected reportPeriodJitter = 15 * 1000;
+
+    constructor(client: Client4, store: Store<GlobalState>) {
+        this.client = client;
+        this.store = store;
+
+        this.counters = new Map();
+        this.histogramMeasures = [];
+
+        // This uses a PerformanceObserver to listen for calls to Performance.measure made by frontend code. It's
+        // recommended to use an observer rather than to call Performance.getEntriesByName directly
+        this.observer = new PerformanceObserver((entries) => this.handleObservations(entries));
+    }
+
+    public observe() {
+        const observedEntryTypes = ['mark', 'measure'];
+        if (PerformanceObserver.supportedEntryTypes.includes('longtask')) {
+            observedEntryTypes.push('longtask');
+        }
+
+        this.observer.observe({
+            entryTypes: observedEntryTypes,
+        });
+
+        // Register handlers for standard metrics and Web Vitals
+        onCLS((metric) => this.handleWebVital(metric));
+        onFCP((metric) => this.handleWebVital(metric));
+        onINP((metric) => this.handleWebVital(metric));
+        onLCP((metric) => this.handleWebVital(metric));
+        onTTFB((metric) => this.handleWebVital(metric));
+
+        // Periodically send performance telemetry to the server, roughly every minute but with some randomness to
+        // avoid overloading the server every minute.
+        this.reportTimeout = window.setTimeout(() => this.handleReportTimeout(), this.nextTimeout());
```

Comment:
this is creating settimeout each minute? Also why intervals wasnt used which is more fitting to do tasks which requires regular interval?

Reply:
I didn't use an interval because I didn't want this to send a report consistently every minute because I was worried that we'd end up in a situation where server CPU usage spikes every 60 seconds because every client sends its performance reports at the same time. Instead, it sends each report between 45 seconds and 75 seconds from the last to prevent all the reports from lining up consistently.

I don't think it's possible to do that with a `setInterval`, so I used a `setTimeout` instead

Reply:
I agree that achieving this with setInterval isn't feasible, but the functionality you're aiming for aligns with what setInterval is designed to accomplish. Regarding the issue of ensuring a 60-second interval for all clients, could we set the interval to a random value between 45 and 75 seconds? This approach should prevent multiple clients from having the same cycle duration. What i am worried about is setTimeouts creation and clearing every 1 minute! With setInterval we dont have to worry about that.